[bigassfan] Fixes and new functionality

mhilbush opened the pull request:
Signed-off-by: Mark Hilbush <mark@hilbush.com>

Fixes and new functionality for BigAssFan fans and wall controllers
- Add support for Wintermode
- Add minspeed and maxspeed to Smarter Cooling mode
- Add initial support for wall controllers
- Add light present channel
- Add periodic polling of fan and controller status
- Fix discovery of H-Series fans
- Fix fan direction channel
- Use ESH ThreadPoolManager

martinvw commented on addons/binding/org.openhab.binding.bigassfan/ESH-INF/thing/thing-types.xml:
```diff
@@ -98,17 +129,32 @@
 		</state>
 	</channel-type>
-	<channel-type id="fan-speed-min">
+	<channel-type id="fan-learn-speed-min">
```
Is learn == "Smart Cooling mode"?

mhilbush replied:
Yes, that's the terminology they use to describe setting the min and max speed when using Smarter Cooling.

martinvw commented on addons/binding/org.openhab.binding.bigassfan/src/main/java/org/openhab/binding/bigassfan/discovery/BigAssFanDiscoveryService.java:
```diff
@@ -182,25 +182,39 @@ private void processMessage(BigAssFanDevice device) {
+                case 3:
+                    // H-Series fans
+                    device.setType(modelParts[0]);
+                    device.setModel(modelParts[2]);
```
What is in modelParts[1], just curious?

mhilbush replied:
Their protocol is driving me nuts...  :-(
```
(Office Fan;DEVICE;ID;20:F8:5E:E8:EE:DE;FAN,LSERIES)
(Master Bedroom Fan;DEVICE;ID;20:F8:5E:E6:47:0B;FAN,HAIKU,HSERIES)
```

martinvw commented on addons/binding/org.openhab.binding.bigassfan/src/main/java/org/openhab/binding/bigassfan/internal/BigAssFanHandlerFactory.java:
```diff
@@ -23,6 +26,13 @@
+    private Logger logger = LoggerFactory.getLogger(BigAssFanHandlerFactory.class);
+
+    @Override
+    protected void activate(ComponentContext componentContext) {
```
@kaikreuzer should we want this?

mhilbush replied:
It's the only way I've found to get the binding version number so that I can log it at startup.  Would be nice if there was a framework call to get the binding version.  Of course, maybe there is already and I just haven't found it.  ;-)

JueBag replied:
I used the following to get the version in a handler:
        `Version version = FrameworkUtil.getBundle(this.getClass()).getVersion();`

mhilbush replied:
Thanks, @JueBag!

@martinvw, if the issue is the reference to bundleContext, I can replace that with the above call.  Just let me know which way to go.

martinvw replied:
Its more about:

> Would be nice if there was a framework call to get the binding version. Of course, maybe there is already and I just haven't found it. ;-)

martinvw replied:
Actually I mean if there is something more generic to get the binding version for the end-users. I don't know whether we should do it like this or that we should just let the users depend on Karaf. But let that be up to @kaikreuzer, @JueBag solution is already a step forward.

mhilbush replied:
If that's the case, in the interest of expediency, I can remove that from my PR until such time as there is a framework call to get the binding version.  :-)

mhilbush commented:
@martinvw I removed the code that gets the version number.  Hopefully, this can move ahead now.

mhilbush commented:
@martinvw  Having removed the code that gets the version number, can this PR move forward now? Thanks!

martinvw commented on addons/binding/org.openhab.binding.bigassfan/src/main/java/org/openhab/binding/bigassfan/handler/BigAssFanHandler.java:
```diff
@@ -405,19 +471,26 @@ private PercentType convertLevelToPercent(String level) {
     }
 
+    private boolean isLightPresent() {
+        if ("PRESENT".equals(fanStateMap.get(CHANNEL_LIGHT_PRESENT).toString())) {
```
I would say that this comparison is fishy, what is the actual type you get from the fanStateMap, calling `toString` on something to compare the string representation is most of the time not the most stable solution.

Also please return a Boolean directly instead of feeding it to an if to convert it to a boolean 🙂

martinvw commented on addons/binding/org.openhab.binding.bigassfan/src/main/java/org/openhab/binding/bigassfan/handler/BigAssFanHandler.java:
```diff
@@ -429,7 +502,7 @@ private void updateChannel(String channelName, State state) {
 
     private String thingID() {
-        return thing.getUID().getId();
+        return thing.getUID().toString();
```
This does not look well, are you sure you want to do this? Where do you use it for.

martinvw replied:
The most places I quickly saw would also be fine if you would just pass `getUID()`, it will be auto converted for both loggers and string concatenation.

martinvw commented on addons/binding/org.openhab.binding.bigassfan/src/main/java/org/openhab/binding/bigassfan/handler/BigAssFanHandler.java:
```diff
@@ -681,167 +770,211 @@ private boolean isMe(String idFromDevice) {
+                return;
+            }
+            logger.debug("Process fan wintermode update for {}: {}", thingID(), messageParts[3]);
+            OnOffType state = messageParts[3].toUpperCase().equals("ON") ? OnOffType.ON : OnOffType.OFF;
```
If you would put the literal left and use equalsIgnoreCase your code would be a little bit more resilient.

martinvw commented on addons/binding/org.openhab.binding.bigassfan/META-INF/MANIFEST.MF:
```diff
@@ -23,6 +23,7 @@ Import-Package: com.google.common.collect,
  org.openhab.binding.bigassfan,
  org.openhab.binding.bigassfan.handler,
  org.osgi.framework,
+ org.osgi.service.component,
```
Do you still need this change or was that for the piece of cod you removed?

martinvw commented on addons/binding/org.openhab.binding.bigassfan/ESH-INF/thing/thing-types.xml:
```diff
@@ -12,12 +12,15 @@
+			<channel id="fan-learn-speed-min" typeId="fan-learn-speed-min" />
+			<channel id="fan-learn-speed-max" typeId="fan-learn-speed-max" />
 			<channel id="fan-speed-min" typeId="fan-speed-min" />
-			<channel id="fan-speed-max" typeId="fan-speed-min" />
```
This looks not backward compatible because fan-speed-min/max suddenly get another meaning, should we publish in the release note to inform users?

mhilbush replied:
It wouldn't be backward compatible if it had actually worked.  :-(  One of the downsides of having to reverse engineer the protocol...

To the best of my knowledge, there's only one other user of the binding, and he wasn't using this feature (because it didn't work).

I don't think we need to add a release note for this.

mhilbush commented:
> Sorry, I'm out of the country right now, so less available

> @kaikreuzer if you have time feel free to take it over

@martinvw @kaikreuzer  No worries.  I'll be unavailable for the next 4-5 days.  I'll get to these when I get back.

mhilbush commented:
@martinvw  I finally got some time to work on this.  Sometimes I think I had more free time when I had a job.  :-o

I believe I addressed all your review comments, except the last one, which I commented on above.

martinvw commented:
PS could you (next time) create a more descriptive commit message. The current one might not mean much to most people. If you would have mentioned the information you added to the PR it would have been more helpful when viewing git history.

mhilbush commented:
> create a more descriptive commit message

Sure.  And thanks for the quick turnaround!
